### PR TITLE
SI-6355, weakend implementation restriction on applyDynamic.

### PR DESCRIPTION
**Pull request by paulp**
I realized one can successfully call an overloaded applyDynamic,
under conditions such as these:

```
  def applyDynamic[T1](m: String)(x1: T1): Any = 1
  def applyDynamic[T1, T2](m: String)(x: T1, y: T2): Any = 2
  def applyDynamic[T1, T2, T3](m: String)(x: T1, y: T2, z: T3): Any = 3
```

So I weakened the overloading restriction to allow overloading
if each method has a distinct number of type parameters. This very
likely still allows the creation of uncallable overloads, but an
overly restrictive rule is worse. If the overload cannot be called,
it will still be discovered at the call site.

### CODE REVIEW

**Comment by JamesIry** [comment]
Code wise LGTM, but the overly restrictive rule has the advantage of being simple to state. Not insisting that this not go in, just less than convinced the extra freedom is ever useful enough to justify the etra cognitive burden for the user.

**Comment by paulp** [comment]
In combination with macros I believe i can use this to write a completely general Dynamic which is not possible with the restriction in place.

**Comment by retronym** [comment]
@lrytz The `git describe` that we added to the build script wasn't quite right; can you change it to something like:

```
git log --oneline head~1...
487380e SI-6146 More accurate prefixes for sealed subtypes.
```

**Comment by lrytz** [comment]
sure, i changed it.

**Comment by JamesIry** [comment]
@paulp In that case maybe reword the error a bit.

```
implementation restriction: applyDynamic cannot be overloaded except by methods with different numbers of type parameters, e.g. applyDynamic[T1](method: String)(arg: T1) and applyDynamic[T1, T2](method: String)(arg1: T1, arg2: T2).
```

WDYT?

**Comment by JamesIry** [comment]
ping @paulp

**Comment by paulp** [comment]
I have acceded to your demands. Let my people go!
